app/src/org/commcare/activities/ReminderThread.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Vector;
+
+import org.commcare.CommCareApplication;
+import org.commcare.suite.model.Alert;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.model.xform.XPathReference;
+import org.javarosa.xpath.XPathParseTool;
+import org.javarosa.xpath.expr.XPathExpression;
+
+import android.content.Context;
+import android.database.Cursor;
+import android.media.Ringtone;
+import android.media.RingtoneManager;
+import android.net.Uri;
+import android.util.Log;
+
+/**
+ * @author ctsims, Saumya Jain
+ *
+ */
+public class ReminderThread {
+    Runnable runnable;
+    Thread myThread;
+    
+    HashMap<String, String> keyToValue;
+    HashMap<String, String> valueToKey;
+
+    private String caseTypeString = "";
+    private String xPathConditionString;
+
+    Context mContext;
+    
+    Cursor c;
+    
+    private boolean mContinuePolling = true;
+    private int sleepTime = 10;
+    
+    public ReminderThread(Context context) {
+        valueToKey = new HashMap<>();
+        keyToValue = new HashMap<>();
+        
+        this.mContext = context;
+        
+        runnable = new Runnable() {
+            @Override
+            public void run() {
+
+                //Prepare xpath expression
+                XPathExpression condition;
+                try{
+                    condition = XPathParseTool.parseXPath(xPathConditionString);
+                }catch (Exception e){
+                    e.printStackTrace();
+                    return;
+                }
+
+                //Prepare evaluation context
+                Hashtable<String, DataInstance> instances = new Hashtable<>();
+                instances.put("casedb", new ExternalDataInstance("jr://instance/casedb","casedb"));
+                EvaluationContext ec = CommCareApplication._().getCurrentSessionWrapper().getEvaluationContext(instances);
+
+                //Prepare TreeReferences
+                String caseXPathRef = "instance('casedb')/casedb/case[@case_type='" + caseTypeString + "']";
+                TreeReference caseType = XPathReference.getPathExpr(caseXPathRef).getReference(); //This is more of an abstract reference to a bunch of stuff in the tree
+                List<TreeReference> refs = ec.expandReference(caseType); //When we expand the reference within the context ec we get the x number of references that actually "exist"
+
+                while(mContinuePolling) {
+                    long current = System.currentTimeMillis();
+
+                    //Evaluate condition
+                    for(TreeReference ref : refs) {
+                        EvaluationContext internal = new EvaluationContext(ec, ref);
+                        Log.d("Result", condition.eval(internal).toString());
```

Comment:
For runtime considerations, I'd cache the results of "condition.eval()" into an object, since you do it again right below and those evaluations could take a considerable amonut of time to evalute